### PR TITLE
Bugs fixing

### PR DESCRIPTION
**Pull request by alexandr-senyuk**
* fix joining to room if user had already been joined before
* fix reconnection
* get rid redundant join event
* use same event names for emitting and listening
* add actions layer
* add middlewares layer

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -15,70 +14,44 @@ const {
...
-            user,
-            room,
-        }
-
-        socket.join(room)
-        socket.emit('message', generateWelcome())
-        socket.broadcast.to(room).emit('message', generateJoined(user))
+    socket.join(context.room)
+    socket.emit('message', generateWelcome())
+    socket.to(context.room).emit('message', generateJoined(user))
```
Comment (sychova) [inline src/index.js]: Why did you remove `broadcast`?

File: src/index.js [hunk 2]
```diff
@@ -15,70 +14,44 @@ const {
...
-const { Location } = require('./entities')
+const actions = require('./actions')
+const middlewares = require('./middlewares')
 
 const app = express()
 const server = http.createServer(app)
 const io = socketio(server)
 
-const publicDirPath = path.join(__dirname, '../public')
+app.use(middlewares)
```
Comment (sychova) [inline src/index.js:26]: Is it some sort of a `reserved` word ('middlwares') for app to use? Or does it just use whatever is exported in that file?

File: src/index.js [hunk 3]
```diff
@@ -17,15 +17,14 @@ const {
...
 
 const app = express()
 const server = http.createServer(app)
 const io = socketio(server)
 
 app.use(middlewares)
-
-app.get('/rooms', actions.getRooms)
-app.get('/room/:roomId/messages', actions.getRoomMessages)
+applyRoutes(app, actions)
```
Comment (sychova) [inline src/index.js:27]: Is it some sort of a reserved word for app to use? Or does it just use whatever is exported in that file?

File: public/js/chat.js [hunk 4]
```diff
@@ -80,6 +73,17 @@ const autoscroll = () => {
     }
 }
 
+const { username, room, roomId } = Qs.parse(location.search, {
+    ignoreQueryPrefix: true,
+})
+
+const socket = io({
+    query: {
```
Comment (sychova) [inline public/js/chat.js:81]: How does the data get here?

File: src/index.js [hunk 5]
```diff
@@ -37,51 +36,47 @@ io.on('connection', async (socket) => {
...
-            user: context.user,
-            room: context.room,
-            message,
+    socket.on('message', async (text, callback) => {
+        const message = await messageToRoomSender.call({
+            ...context,
+            message: text,
             type: 'text',
         })
         io.to(context.room).emit('message', {
```
Comment (sychova) [inline src/index.js:58]: How does this work?
I mean that `on('message')` a `emit('message')` happens? O_o
Or does it mean that we can't 'handle' the `emit` in the same file with an `on` action? O_O

File: src/index.js [hunk 6]
```diff
@@ -37,51 +36,47 @@ io.on('connection', async (socket) => {
...
-    socket.join(context.room)
+    const context = { user, room }
+    socket.join(room)
     socket.emit('message', generateWelcome())
-    socket.to(context.room).emit('message', generateJoined(user))
+    socket.to(room).emit('message', generateJoined(user))
 
-    const roomWithUsers = await roomUsersFetcher.call(room)
-    io.to(room).emit('roomData', roomWithUsers)
+    const syncRoomData = async () => {
```
Comment (sychova) [inline src/index.js:44]: Why is this method not cut out into own service/something like others?
Why is it defined here?

Comment (alexandr-senyuk) [reply src/index.js:44]: It could be segregated. But it is depends on data transfer layer

File: public/js/chat.js (hunk 4 above)
Comment (alexandr-senyuk) [reply public/js/chat.js:81]: Which data? query ? it is the same as before - from url

File: src/index.js (hunk 3 above)
Comment (alexandr-senyuk) [reply src/index.js:27]: actions? this is kinda part of MVC pattern

File: src/index.js (hunk 2 above)
Comment (alexandr-senyuk) [reply src/index.js:26]: it is kinda pattern. And it is a part of express 
https://expressjs.com/en/guide/using-middleware.html

File: src/index.js (hunk 1 above)
Comment (alexandr-senyuk) [reply src/index.js]: https://socket.io/docs/v4/emit-cheatsheet
it is redundant

File: src/index.js (hunk 5 above)
Comment (alexandr-senyuk) [reply src/index.js:58]: i didn't get you

File: src/index.js (hunk 1 above)
Comment (sychova) [reply src/index.js]: But it's there. In the latest version. O_o

File: public/js/chat.js (hunk 4 above)
Comment (sychova) [reply public/js/chat.js:81]: But the old code didn't pass anything to `io`. What does the passing mean?

Comment (sychova) [reply public/js/chat.js:81]: But the old code didn't pass anything to `io`. What does the passing mean?

File: src/index.js (hunk 5 above)
Comment (sychova) [reply src/index.js:58]: Inside a 'reaction' to an event you emit the same event you're 'reacting' to.
